xclim/indices/_agro.py

```diff
@@ -882,6 +883,188 @@ def water_budget(
+    thresh_dry_end = convert_units_to(thresh_dry_end, pram)
+
+    last_doy = pram.indexes["time"][-1].strftime("%m-%d")
+
+    # Eliminate negative values.
+    pram = xarray.where(pram < 0, 0, pram)
+    pram.attrs["units"] = "mm"
+
+    # should we flag date_min_end  < date_max_start?
+    def _get_first_run(run_positions, start_date, end_date):
+        run_positions = select_time(run_positions, date_bounds=(start_date, end_date))
+        first_start = run_positions.argmax("time")
+        return xarray.where(
+            first_start != run_positions.argmin("time"), first_start, np.NaN
+        )
+
+    # Find the start of the rain season
+    def _get_first_run_start(pram, window_dry=window_dry_start):
+        pram = select_time(pram, date_bounds=(date_min_start, last_doy))
+
+        # First condition: Start with enough precipitation
+        da_start = pram.rolling({"time": window_wet_start}).sum() >= thresh_wet_start
+
+        # Second condition: No dry period after
+        if method_dry_start == "per_day":
+            da_stop = pram <= thresh_dry_start
+        elif method_dry_start == "total":
+            da_stop = pram.rolling({"time": window_dry}).sum() <= thresh_dry_start
+            # equivalent to rolling forward in time instead, i.e. end date will be at beginning of dry run
+            da_stop = da_stop.shift({"time": -(window_dry - 1)}, fill_value=False)
+            window_dry = 1
+
+        # First and second condition combined in a run length
+        run_positions = (
+            rl.rle_events(da_start, 1, da_stop, window_dry, "first")
+            >= window_not_dry_start + window_wet_start
+        )
+
+        return _get_first_run(run_positions, date_min_start, date_max_start)
+
+    # Find the end of the rain season
+    def _get_first_run_end(pram):
+        if method_dry_end == "per_day":
+            da_stop = pram <= thresh_dry_end
+            run_positions = rl.rle(da_stop, index="first") >= window_dry_end
+        elif method_dry_end == "total":
+            da_stop = pram.rolling({"time": window_dry_end}).sum() <= thresh_dry_end
+            run_positions = da_stop
+        return _get_first_run(run_positions, date_min_end, date_max_end)
+
+    # Get start, end and length of rain season. Written as a function so it can be resampled
+    def _get_rain_season(pram):
+        start = _get_first_run_start(pram)
+
+        # masking value before  start of the season (end of season should be after)
+        start_ind = xarray.where(start.notnull(), start.astype(int), -1)
+        mask = xarray.where(
+            pram.time == pram.isel(time=start_ind).time, 1, np.NaN
+        ).ffill("time")
+        mask[{"time": start_ind}] = np.NaN
+        mask = mask.notnull()
+        end = _get_first_run_end(pram.where(mask))
+
+        length = xarray.where(end.notnull(), end - start, pram["time"].size - start)
+
+        # converting to doy
+        crd = pram["time"]
+        crd = getattr(crd.dt, "dayofyear")
```

Comment:
Héhé copied from `run_length.py` :) ?
```suggestion
        crd = pram.time.dt.dayofyear
```

Reply:
mayyybe